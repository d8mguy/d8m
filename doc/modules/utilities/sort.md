## Sort

Sort provides sorting of arbitrary types, both generically and by instantiation. The generic sort works by arranging to call the go sort package. Thus, standard calls to `modsort` and `sort` instantiate a sort algorithm for the particular element type of the list. This code is lifted very directly from the go package code except that type specific methods are used instead of generic ones. Instantiated sorting is quite a bit faster than generic; you can find details in the performance data section.

The advantage of generic sorting is that when sorting small lists, the asymptotic speedup doesn't matter, and each distinct instantiation of the full sorting algorithm results in a non-trivial amount of code, so significant bloat is possible if you're sorting small lists of many different element types. To get generic sorting, you apply a special type generator, called `GSLT` ("generically sortable list type"), that creates list types augmented with the methods required by the go sort package. The sort module defines and enables a rewrite rule that changes calls to `modsort` on list types created by `GSLT` into calls to `Sort` in the go package. (Note that `sort.Sort` sorts its argument in place, just like `modsort`.)

#### modsort: function

    val modsort : \mod(list(OT)) -> list(OT) for any OT::ordered
Modsort is the main entry point for sorting. This version is for lists of ordered elements. Modsort simply calls quickSort.

    val modsort : \mod(list(T), fn:\(T,T)->boolean) -> list(T) for any T::entity
Modsort is the main entry point for sorting. This version is for arbitrary elements, plus a function to compare the elements. Modsort simply calls quickSort.

    val modsort : \mod(list(T), fn:\(T)->OT) -> list(T) for any T::entity
Modsort is the main entry point for sorting. This version is for arbitrary elements, plus a function to extract an ordered element from a list element. Unlike with olist, there is no need for the extracted values to form a total order with resepect to the collection. Modsort simply calls quickSort.

#### sort: function

    val sort : \(list(OT)) -> list(OT) for any OT::ordered
    val sort : \(list(T), fn:\(T,T)->boolean) -> list(T) for any T::entity
    val sort : \(list(T), fn:\(T)->OT) -> list(T) for any T::entity

Sort provides a "pure" version of modsort by making a shallow copy of its argument, calling modsort on it, and returning it. 

#### quickSort: function

    val quickSort : \mod(list(OT), first, last, mxdepth: integer) -> nothing for any OT::ordered
    val quickSort : \mod(list(T), fn:\(T,T)->boolean, first, last, mxdepth: integer) -> nothing for any T::entity

QuickSort sorts `lst[first...last]`, switching between heapSort, insertionSort, and quickSort algorithms based on the size of the sorting interval defined by `last - first`.

#### insertionSort: function

    val insertionSort = \mod(lst:list(OT), first, last: integer) -> nothing {...} for any OT::ordered
    val insertionSort = \mod(lst:list(T), fn:\(T,T)->boolean, first, last: integer) -> nothing {...} for any T::entity

InsertionSort implements the insertion sort algorithm on `lst[first...last]`. It's exported in case you want to use it for something but is mainly used by quickSort.

#### heapSort: function

    val heapSort = \mod(lst:list(OT), first, last: integer) -> nothing {...} for any OT::ordered
    val heapSort = \mod(lst:list(T), fn:\(T,T)->boolean, first, last: integer) -> nothing {...} for any T::entity

HeapSort implements the heap sort algorithm on `lst[first...last]`. It's exported in case you want to use it for something but is mainly used by quickSort.

#### inOrder: function

    val inOrder = \(lst:list(OT)) -> boolean {...} for any OT::ordered

InOrder returns whether the elements of lst are in order (according to `OT`'s ordering).

#### GSLT: typegen

    val GSLT = \typegen(OT::ordered) -> type {...}

GSLT returns a `list(OT)` augmented with methods needed by go's sort package for sorting types generically. The declaration provides effective assertions that ensure correct code generation for such calls. The sort module also defines and activates a rewrite rule that rewrites a call to `modsort` on any GSLT into a call to sort.Sort (in the go package). Therefore, you can get a list sorted generically by creating it as a GSLT, then calling modsort on it. Since it's otherwise a standard list, you can use it in all the usual ways outside of sorting.
