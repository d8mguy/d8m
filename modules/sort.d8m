// Sort provides sorting of arbitrary types, both generically and by instantiation.
// The generic sort works by arranging to call the go sort package. Thus, standard calls to `modsort` and `sort`
// instantiate a sort algorithm for the particular element type of the list. This code is lifted very directly
// from the go package code except that type specific methods are used instead of generic ones. Instantiated sorting
// is quite a bit faster than generic; you can find details in the performance data section.
//
// The advantage of generic sorting is that when sorting small lists, the asymptotic speedup doesn't matter,
// and each distinct instantiation of the full sorting algorithm results in a non-trivial amount of code,
// so significant bloat is possible if you're sorting small lists of many different element types.
// To get generic sorting, you apply a special type generator, called `GSLT` ("generically sortable list type"),
// that creates list types augmented with the methods required by the go sort package. The sort module defines
// and enables a rewrite rule that changes calls to `modsort` on list types created by `GSLT` into calls to `Sort`
// in the go package. (Note that `sort.Sort` sorts its argument in place, just like `modsort`.)

export modsort, sort, quickSort, insertionSort, heapSort, maxDepth, inOrder, GSLT

// 2*ceiling(lg(n+1)), used to decide when to switch between heapsort and quicksort.
val maxDepth = \imp(n: integer) {
	var depth: integer = 0, i: integer = n
	while(i > 0) { depth += 1; i /= 2 }
	depth * 2
}

// First, define the panoply of sorting ftns on (lists of) ordered entities
given(OT::ordered) {
	// InsertionSort implements the insertion sort algorithm on `lst[first...last]`. It's exported
	// in case you want to use it for something but is mainly used by quickSort.
	val insertionSort = \mod(coll: list(OT), first, last: integer) {
		val firstp1 = first+1
		each(i^(firstp1...last), j^reverse(firstp1..i)) if(coll[j] < coll[j-1]) coll.swap(j, j-1) else break
	}
	val siftDown = \mod(coll: list(OT), lo, hi, first: integer) {
		var root = lo
		loop {
			var child = 2*root + 1
			if(child >= hi) break
			if(child+1 < hi && coll[first+child] < coll[first+child+1]) child += 1
			unless(coll[first+root] < coll[first+child]) return
			coll.swap(first+root, first+child)
			root = child
		}
	}
	
	// HeapSort implements the heap sort algorithm on `lst[first...last]`. It's exported in case you want to use
	// it for something but is mainly used by quickSort.
	val heapSort = \mod(coll: list(OT), first, last: integer) {
		val len = last - first
		each(i^reverse(0..(len-1)/2)) siftDown(coll, i, len, first)
		each(i^reverse(0..(len-1))) { coll.swap(first, first+i); siftDown(coll, 0, i, first) }
	}
	
	// sort the elts at m1, m0, and m2
	val medianOf3 = \mod(coll: list(OT), m1, m0, m2: integer) {
		if(coll[m1] < coll[m0]) { coll.swap(m1, m0) }
		if(coll[m2] < coll[m1]) { coll.swap(m2, m1);  if(coll[m1] < coll[m0]) coll.swap(m1, m0) }
	}
	
	val doPivot = \mod(coll: list(OT), lo, hi: integer) {
		var m = (lo+hi)/2
		if(hi - lo > 40) {
			var s = (hi - lo)/8
			medianOf3(coll, lo, lo+s, lo+2*s)
			medianOf3(coll, m, m-s, m+s)
			medianOf3(coll, hi-1, hi-1-s, hi-1-2*s)
		}
		medianOf3(coll, lo, m, hi-1)
		var pivot = lo, a = lo+1, c = hi-1
		while(a < c && coll[a] < coll[pivot]) a += 1
		var b = a
		loop {
			while(b < c && !(coll[pivot] < coll[b])) b += 1
			while(b < c && coll[pivot] < coll[c-1]) c -= 1
			if(b >= c) break
			coll.swap(b, c-1)
			b += 1
			c -= 1
		}		
		var protect = hi - c < 5
		if(!protect && hi-c < (hi-lo)/4) {
			var dups = 0
			if(coll[hi-1] <= coll[pivot]) { coll.swap(c, hi-1); c += 1; dups += 1 }
			if(coll[pivot] <= coll[b-1]) { b -= 1; dups += 1 }
			if(coll[pivot] <= coll[m]) { coll.swap(m, b-1); b -= 1; dups += 1 }
			protect = dups > 1
		}
		if(protect) {
			loop {
				while(a < b && !(coll[b-1] < coll[pivot])) b -= 1
				while(a < b && coll[a] < coll[pivot]) a += 1
				if(a >= b) break
				coll.swap(a, b-1)
				a += 1
				b -= 1
			}
		}
		coll.swap(pivot, b-1)
		[mlo~b-1, mhi~c]
	}
	
	// QuickSort sorts `lst[first...last]`, switching between heapSort, insertionSort, and quickSort algorithms
	// based on the size of the sorting interval defined by `last - first`.
	val quickSort = \mod(coll: list(OT), first0, last0, mxdepth: integer) -> nothing {
		var mxd = mxdepth
		var first = first0, last = last0
		while(last - first > 12) {
			if(mxd == 0) { heapSort(coll, first, last); return }
			mxd -= 1
			val tmp = doPivot(coll, first, last)
			if(tmp.mlo - first < last - tmp.mhi) { quickSort(coll, first, tmp.mlo, mxd); first = tmp.mhi }
			else { quickSort(coll, tmp.mhi, last, mxd); last = tmp.mlo }
		}
		if(last - first > 1) {		// interval size in range 2..12
			var i = first + 6
			while(i < last) { if(coll[i] < coll[i-6]) coll.swap(i, i-6); i += 1 }
			insertionSort(coll, first, last)
		}
	}
	
	// Modsort is the main entry point for sorting. This version is for lists of ordered elements. Modsort simply calls quickSort.
	val modsort = \mod(coll: list(OT)) -> list(OT) {
		quickSort(coll, 0, coll.count, maxDepth(coll.count))
		coll
	}
	// Sort provides a "pure" version of modsort by making a shallow copy of its argument, calling modsort on it, and returning it.
	val sort = \imp(coll: list(OT)) {
		var ret = copy(coll)
		modsort(ret) 
	}
}

// Second, define the same sorting ftns on (lists of) general entities, adding a comparison ftn to the args.
given(T::entity) {
	val insertionSort = \mod(coll: list(T), fn:\(T,T)->boolean, first, last: integer) {
		val firstp1 = first+1
		each(i^(firstp1...last), j^(firstp1..i).reverse) if(fn(coll[j], coll[j-1])) coll.swap(j, j-1) else break
	}
	val siftDown = \mod(coll: list(T), fn:\(T,T)->boolean, lo, hi, first: integer) {
		var root = lo
		loop {
			var child = 2*root + 1
			if(child >= hi) break
			if(child+1 < hi && fn(coll[first+child], coll[first+child+1])) child += 1
			unless(fn(coll[first+root], coll[first+child])) return
			coll.swap(first+root, first+child)
			root = child
		}
	}
	val heapSort = \mod(coll: list(T), fn:\(T,T)->boolean, first, last: integer) {
		val len = last - first
		each(i^reverse(0..(len-1)/2)) siftDown(coll, fn, i, len, first)
		each(i^reverse(0..(len-1))) { coll.swap(first, first+i); siftDown(coll, fn, 0, i, first) }
	}
	val medianOf3 = \mod(coll: list(T), fn:\(T,T)->boolean, m1, m0, m2: integer) {
		if(fn(coll[m1], coll[m0])) { coll.swap(m1, m0) }
		if(fn(coll[m2], coll[m1])) { coll.swap(m2, m1);  if(fn(coll[m1], coll[m0])) coll.swap(m1, m0) }
	}
	val doPivot = \mod(coll: list(T), fn:\(T,T)->boolean, lo, hi: integer) {
		var m = (lo+hi)/2
		if(hi - lo > 40) {
			var s = (hi - lo)/8
			medianOf3(coll, fn, lo, lo+s, lo+2*s)
			medianOf3(coll, fn, m, m-s, m+s)
			medianOf3(coll, fn, hi-1, hi-1-s, hi-1-2*s)
		}
		medianOf3(coll, fn, lo, m, hi-1)
		var pivot = lo, a = lo+1, c = hi-1
		while(a < c && fn(coll[a], coll[pivot])) a += 1
		var b = a
		loop {
			while(b < c && !fn(coll[pivot], coll[b])) b += 1
			while(b < c && fn(coll[pivot], coll[c-1])) c -= 1
			if(b >= c) break
			coll.swap(b, c-1)
			b += 1
			c -= 1
		}		
		var protect = hi - c < 5
		if(!protect && hi-c < (hi-lo)/4) {
			var dups = 0
			if(!fn(coll[pivot], coll[hi-1])) { coll.swap(c, hi-1); c += 1; dups += 1 } 
			if(!fn(coll[b-1], coll[pivot])) { b -= 1; dups += 1 } 
			if(!fn(coll[m], coll[pivot])) { coll.swap(m, b-1); b -= 1; dups += 1 }
			protect = dups > 1
		}
		if(protect) {
			loop {
				while(a < b && !fn(coll[b-1], coll[pivot])) b -= 1
				while(a < b && fn(coll[a], coll[pivot])) a += 1
				if(a >= b) break
				coll.swap(a, b-1)
				a += 1
				b -= 1
			}
		}
		coll.swap(pivot, b-1)
		[mlo~b-1, mhi~c]
	}
	
	val quickSort = \mod(coll: list(T), fn:\(T,T)->boolean, first0, last0, mxdepth: integer) -> nothing {
		var mxd = mxdepth
		var first = first0, last = last0
		while(last - first > 12) {
			if(mxd == 0) { heapSort(coll, fn, first, last); return }
			mxd -= 1
			val tmp = doPivot(coll, fn, first, last)
			if(tmp.mlo - first < last - tmp.mhi) { quickSort(coll, fn, first, tmp.mlo, mxd); first = tmp.mhi }
			else { quickSort(coll, fn, tmp.mhi, last, mxd); last = tmp.mlo }
		}
		if(last - first > 1) {		// interval size in range 2..12
			var i = first + 6
			while(i < last) { if(fn(coll[i], coll[i-6])) coll.swap(i, i-6); i += 1 }
			insertionSort(coll, fn, first, last)
		}
	}
	
	// Modsort is the main entry point for sorting. This version is for arbitrary elements, plus a
	// function to compare the elements. Modsort simply calls quickSort.
	val modsort = \mod(coll: list(T), fn:\(T,T)->boolean) {
		quickSort(coll, fn, 0, coll.count, maxDepth(coll.count))
		coll
	}
	val sort = \imp(coll: list(T), fn:\(T,T)->boolean) {
		var ret = copy(coll)
		modsort(ret, fn) 
	}
	
}

// For sorting general entities with an order extraction ftn, build in the extract and sort optimzn,
// plus the sort in place one once we have permuted indices.
given(T::entity, OT::ordered) {
	val modsort = \mod(lst: list(T), extract:\(T)->OT) {
		val XT = tuple(xtd: OT, inx: integer)
		var pairs = modsort(lst.{ [XT: extract(this), index] }, \(x,y: XT) { x.xtd < y.xtd })
		// now we have the permutation in pairs.{inx}; use perm->cycles algo to move in place
		// find cycle starting at 0, moving as we go, and marking moved indices by setting pairs.inx to -1.
		var topinx = 0, curinx = 0
		loop {
            curinx = topinx
            val tmpval = lst[curinx]
		    loop {
		        val nxinx = pairs[curinx].inx
		        pairs[curinx].inx = -1          // mark as seen
		        if(nxinx == curinx) break        // this only happens with 1-cycles, which require no mvmt
		        if(nxinx == topinx) {
                    lst[curinx] = tmpval
		            break
		        }
		        lst[curinx] = lst[nxinx]
		        curinx = nxinx
		    }
		    // we've completed a cycle; check if there's another
		    while(topinx < lst.count && pairs[topinx].inx < 0) topinx += 1
		    if(topinx == lst.count) break
		}
    	lst
	}
	val sort = \imp(lst: list(T), extract:\(T)->OT) {
		var ret = copy(lst)
		modsort(ret, extract)
	}
}

// answer whether a list(T) is ordered according to fn. 

given(OT::ordered) {
    // InOrder returns whether the elements of lst are in order (according to `OT`'s ordering).
    val inOrder = \(lst: list(OT)) {
        if(lst.count > 1) {
            each(inx^1...lst.count) unless(lst[inx-1] <= lst[inx]) return false
        }
        true
    }
}

// GSLT returns a `list(OT)` augmented with methods needed by go's sort package for sorting types generically.
// The declaration provides effective assertions that ensure correct code generation for such calls.
// The sort module also defines and activates a rewrite rule that rewrites a call to `modsort` on any GSLT into a call to sort.Sort (in the go package).
// Therefore, you can get a list sorted generically by creating it as a GSLT, then calling modsort on it.
// Since it's otherwise a standard list, you can use it in all the usual ways outside of sorting.
val GSLT = \typegen(T::ordered) {
	extend list(T) where {
		method Len = \() { count(self) }
		method Swap = \mod(i, j: integer) { self.swap(i, j) }
		method Less = \(i, j: integer) { self[i] < self[j] }
		assert Gomethod(Len)
		assert Gomethod(Swap)
		assert Gomethod(Less)
	}
}
val GncSortable = typepred(T::list,
	Len:\(T)->integer,
	Swap:\mod(T, integer, integer)->nothing,
	Less:\(T,integer,integer)->boolean)
given(GST::GncSortable) {
	val GncSortRW = \xprftn() {
		val srtble: GST
		val sortstmt: xprterm = matching("modsort(srtble)", false, true)
		matchquery(sortstmt)
		sortstmt.subst([sortstmt], [mkterm(:funcall, pkgSymbol("sort", "Sort"), [srtble])])
	}
}
rewrite(GncSortRW)


