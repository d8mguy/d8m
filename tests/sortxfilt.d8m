// sortxfilt
import "sort" melted
given(T0 :: entity, OT :: ordered) {
    val extractAndSort = \xprftn() {
          val biglist: list(T0), fn:\(T0,T0)->boolean, K : integer
          matchquery("sort(biglist, fn).[index < K]")
          skipIf(family(T0) != :tuple || attributes(T0).count <= 2)
          val A = fn.fmlargs[0], B = fn.fmlargs[1], AT : xprsym
          val relop : xprsym = matching([:<, :<=, :>, :>=].{ methodNamed(OT, this) }, false, false)
          val fnPtn = pattern(mkterm(:funcall, relop, [mkterm(:symchain, A, AT), mkterm(:symchain, B, AT)]))
          skipIf(fn.defn.stmts.count != 1 || !match(fnPtn, fn.defn.stmts[0]))
          val XT = mktype(:tuple, [AT, mksym(:inx, integer)])
          val xfnargs = fn.fmlargs.{ mksym(this.ident, XT) }
          val xfnbody = fn.defn.subst(fn.fmlargs, xfnargs)
          val xfn = mkftn(typesubst(fn.type, T0, XT), xfnargs, xfnbody)
          val output = "sort(map(item^biglist) { [XT: item.AT, index] }, xfn).[index < K].{ biglist[this.inx] }"
          parseRW(output)
    }
}
rewrite(extractAndSort)
import go "math/rand"
val tt = tuple(xx : float, yy : string, zz : integer)
val randx = rand.New(rand.NewSource(1))
val aLst = (1..800).{ val f = randx.Float64(); [tt: f, "xxx", round(5500.0*f)] }
sort(aLst, \(a,b : tt) { a.zz < b.zz }).[index < 20]