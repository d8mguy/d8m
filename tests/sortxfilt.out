[[0.000514; xxx; 3], [0.001412; xxx; 8], [0.001904; xxx; 10], [0.002843; xxx; 16], [0.004445; xxx; 24], [0.004489; xxx; 25], [0.004880; xxx; 27], [0.006627; xxx; 36], [0.007528; xxx; 41], [0.007681; xxx; 42], [0.011610; xxx; 64], [0.012481; xxx; 69], [0.012826; xxx; 71], [0.013286; xxx; 73], [0.014827; xxx; 82], [0.015181; xxx; 83], [0.018404; xxx; 101], [0.018713; xxx; 103], [0.019875; xxx; 109], [0.019809; xxx; 109]]
